### PR TITLE
Restore leak tracking in tests

### LINKED ISSUE #10226: Restore buffer leak tracking in tests.
**Jetty version(s)**
12

**Description**
Build on `ArrayByteBufferPool.Tracking` introduced in #10225 to fix the TODOs named "restore leak tracking".

### PR DESCRIPTION
**Pull request by lorban**
- Restore leak tracking in every test where that makes sense.
- Improve the leak tracking buffer pool to collect more data.
- Add the ability to dump the heap when a leak is detected.
- Add the ability to disable leak tracking for a particular test or an entire test class with `@Tag("DisableLeakTracking")`, only for server or client or only for specific transports.
- Fix some detected leaks.
- Disable leak tracking for the remaining detected leaks.

See #10226.

### CODE REVIEW

File: jetty-core/jetty-fcgi/jetty-fcgi-server/src/test/java/org/eclipse/jetty/fcgi/server/AbstractHttpClientServerTest.java [hunk 1]
```diff
@@ -65,33 +58,28 @@ public void start(Handler handler) throws Exception
...
-        assertThat("Connection Leaks", connectionLeaks.get(), Matchers.is(0L));
-
-        if (client != null)
-            client.stop();
-        if (server != null)
-            server.stop();
+        try
+        {
+            if (serverBufferPool != null)
+                assertThat("Server Leaks: " + serverBufferPool.getLeaks(), serverBufferPool.getLeaks().size(), Matchers.is(0));
```
Comment (sbordet) [inline jetty-core/jetty-fcgi/jetty-fcgi-server/src/test/java/org/eclipse/jetty/fcgi/server/AbstractHttpClientServerTest.java]: Use `dumpLeaks()` to better dump the leaks in case of assertion failure.

File: jetty-core/jetty-fcgi/jetty-fcgi-server/src/test/java/org/eclipse/jetty/fcgi/server/AbstractHttpClientServerTest.java [hunk 2]
```diff
@@ -65,33 +58,28 @@ public void start(Handler handler) throws Exception
...
+            if (clientBufferPool != null)
+                assertThat("Client Leaks: " + clientBufferPool.getLeaks(), clientBufferPool.getLeaks().size(), Matchers.is(0));
```
Comment (sbordet) [inline jetty-core/jetty-fcgi/jetty-fcgi-server/src/test/java/org/eclipse/jetty/fcgi/server/AbstractHttpClientServerTest.java]: Ditto.

File: jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ThreadStarvationTest.java [hunk 3]
```diff
@@ -147,7 +151,15 @@ private Server prepareServer(Scenario scenario, Handler handler)
     @AfterEach
     public void dispose() throws Exception
     {
-        _server.stop();
+        ArrayByteBufferPool.Tracking byteBufferPool = (ArrayByteBufferPool.Tracking)_server.getConnectors()[0].getByteBufferPool();
+        try
+        {
+            assertThat("Server Leaks: " + byteBufferPool.getLeaks(), byteBufferPool.getLeaks().size(), Matchers.is(0));
```
Comment (sbordet) [inline jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ThreadStarvationTest.java]: Ditto

File: jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/DebugHandlerTest.java [hunk 4]
```diff
@@ -130,7 +128,15 @@ public void handle(String target, Request baseRequest, HttpServletRequest reques
     @AfterEach
     public void stopServer() throws Exception
     {
-        server.stop();
+        ArrayByteBufferPool.Tracking byteBufferPool = (ArrayByteBufferPool.Tracking)server.getConnectors()[0].getByteBufferPool();
+        try
+        {
+            assertThat("Server Leaks: " + byteBufferPool.getLeaks(), byteBufferPool.getLeaks().size(), Matchers.is(0));
```
Comment (sbordet) [inline jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/DebugHandlerTest.java]: Ditto

File: jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ssl/ServerConnectorSslServerTest.java [hunk 5]
```diff
@@ -109,6 +107,13 @@ public void init() throws Exception
         }
     }
 
+    @AfterEach
+    public void dispose() throws Exception
+    {
+        ArrayByteBufferPool.Tracking byteBufferPool = (ArrayByteBufferPool.Tracking)_server.getConnectors()[0].getByteBufferPool();
+        assertThat("Server Leaks: " + byteBufferPool.getLeaks(), byteBufferPool.getLeaks().size(), Matchers.is(0));
```
Comment (sbordet) [inline jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ssl/ServerConnectorSslServerTest.java]: Ditto

File: jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/AbstractTest.java [hunk 6]
```diff
@@ -109,8 +113,18 @@ public static Collection<Transport> transportsTCP()
...
     public void dispose()
     {
-        LifeCycle.stop(client);
-        LifeCycle.stop(server);
+        try
+        {
+            if (serverBufferPool != null)
+                assertThat("Server Leaks: " + serverBufferPool.getLeaks(), serverBufferPool.getLeaks().size(), Matchers.is(0));
+            if (clientBufferPool != null)
+                assertThat("Client Leaks: " + clientBufferPool.getLeaks(), clientBufferPool.getLeaks().size(), Matchers.is(0));
```
Comment (sbordet) [inline jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/AbstractTest.java]: Ditto

File: jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/HttpClientLoadTest.java [hunk 7]
```diff
@@ -105,7 +85,7 @@ protected void leaked(LeakDetector<Connection>.LeakInfo leakInfo)
             run(transport, iterations);
         }
 
-        assertLeaks();
+        assertThat("Leaks: " + byteBufferPool.getLeaks(), byteBufferPool.getLeaks().size(), Matchers.is(0));
```
Comment (sbordet) [inline jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/HttpClientLoadTest.java]: Ditto.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-core/jetty-fcgi/jetty-fcgi-server/src/test/java/org/eclipse/jetty/fcgi/server/AbstractHttpClientServerTest.java (hunk 1 above)
Comment (lorban) [reply jetty-core/jetty-fcgi/jetty-fcgi-server/src/test/java/org/eclipse/jetty/fcgi/server/AbstractHttpClientServerTest.java]: Done.

File: jetty-core/jetty-fcgi/jetty-fcgi-server/src/test/java/org/eclipse/jetty/fcgi/server/AbstractHttpClientServerTest.java (hunk 2 above)
Comment (lorban) [reply jetty-core/jetty-fcgi/jetty-fcgi-server/src/test/java/org/eclipse/jetty/fcgi/server/AbstractHttpClientServerTest.java]: Done.

File: jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ThreadStarvationTest.java (hunk 3 above)
Comment (lorban) [reply jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ThreadStarvationTest.java]: Done.

File: jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ssl/ServerConnectorSslServerTest.java (hunk 5 above)
Comment (lorban) [reply jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ssl/ServerConnectorSslServerTest.java]: Done.

File: jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/AbstractTest.java (hunk 6 above)
Comment (lorban) [reply jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/AbstractTest.java]: Done.

File: jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/DebugHandlerTest.java (hunk 4 above)
Comment (lorban) [reply jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/DebugHandlerTest.java]: Done.

File: jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/HttpClientLoadTest.java (hunk 7 above)
Comment (lorban) [reply jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/HttpClientLoadTest.java]: Done.

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java [hunk 8]
```diff
@@ -603,32 +605,34 @@ public Tracking(int minCapacity, int maxCapacity, int maxBucketSize, long maxHea
...
         public String dumpLeaks()
         {
             return getLeaks().stream()
-                .map(Buffer::dump)
+                .map(TrackingBuffer::dump)
                 .collect(Collectors.joining(System.lineSeparator()));
         }
 
-        public class Buffer extends RetainableByteBuffer.Wrapper
+        public class TrackingBuffer extends RetainableByteBuffer.Wrapper
```
Comment (sbordet) [inline jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java]: I still prefer this class just be called `Buffer`, as `Tracking` is already the outer class.
We will eventually improve the toString() and for now we know what DEBUG logging prints.

File: jetty-core/jetty-fcgi/jetty-fcgi-server/pom.xml [hunk 9]
```diff
@@ -22,6 +22,8 @@
           <argLine>
             @{argLine} ${jetty.surefire.argLine}
             --add-reads org.eclipse.jetty.fcgi.server=org.eclipse.jetty.logging
+            --add-reads org.eclipse.jetty.fcgi.server=java.management
+            --add-reads org.eclipse.jetty.fcgi.server=jdk.management
```
Comment (sbordet) [inline jetty-core/jetty-fcgi/jetty-fcgi-server/pom.xml]: Why are these necessary?

File: jetty-core/jetty-fcgi/jetty-fcgi-server/src/test/java/org/eclipse/jetty/fcgi/server/AbstractHttpClientServerTest.java [hunk 10]
```diff
@@ -65,33 +61,46 @@ public void start(Handler handler) throws Exception
...
+            {
+                try
+                {
+                    Awaitility.await().atMost(3, TimeUnit.SECONDS).until(() -> serverBufferPool.getLeaks().size(), Matchers.is(0));
+                }
+                catch (Exception e)
+                {
+                    fail(e.getMessage() + "\n---\nServer Leaks: " + serverBufferPool.dumpLeaks() + "---\n");
+                }
+            }
```
Comment (sbordet) [inline jetty-core/jetty-fcgi/jetty-fcgi-server/src/test/java/org/eclipse/jetty/fcgi/server/AbstractHttpClientServerTest.java]: Make an `assertLeaks()` method to avoid code duplication.

File: jetty-core/jetty-fcgi/jetty-fcgi-server/src/test/java/org/eclipse/jetty/fcgi/server/HttpClientTest.java [hunk 11]
```diff
@@ -240,6 +240,7 @@ public boolean handle(org.eclipse.jetty.server.Request request, org.eclipse.jett
                 String paramValue2 = fields.getValue(paramName2);
                 assertEquals("", paramValue2);
                 Content.Sink.write(response, true, UTF_8.encode("empty"));
+                callback.succeeded();
```
Comment (sbordet) [inline jetty-core/jetty-fcgi/jetty-fcgi-server/src/test/java/org/eclipse/jetty/fcgi/server/HttpClientTest.java]:
```suggestion
                Content.Sink.write(response, true, "empty", callback);
```

File: jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/DebugHandlerTest.java [hunk 12]
```diff
@@ -130,7 +128,15 @@ public void handle(String target, Request baseRequest, HttpServletRequest reques
     @AfterEach
     public void stopServer() throws Exception
     {
-        server.stop();
+        ArrayByteBufferPool.Tracking byteBufferPool = (ArrayByteBufferPool.Tracking)server.getConnectors()[0].getByteBufferPool();
```
Comment (sbordet) [inline jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/DebugHandlerTest.java]: Yuck :)
You are only checking one of the tracking buffer pools.
I'd suggest to save them as fields and use the fields here, rather than scavenging them from the server.

File: jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ssl/ServerConnectorSslServerTest.java [hunk 13]
```diff
@@ -109,6 +107,13 @@ public void init() throws Exception
         }
     }
 
+    @AfterEach
+    public void dispose() throws Exception
+    {
+        ArrayByteBufferPool.Tracking byteBufferPool = (ArrayByteBufferPool.Tracking)_server.getConnectors()[0].getByteBufferPool();
```
Comment (sbordet) [inline jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ssl/ServerConnectorSslServerTest.java]: Better to use a field.

File: jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/AbstractTest.java [hunk 14]
```diff
@@ -107,10 +121,131 @@ public static Collection<Transport> transportsTCP()
...
+                {
+                    Awaitility.await().atMost(3, TimeUnit.SECONDS).until(() -> serverBufferPool.getLeaks().size(), Matchers.is(0));
+                }
+                catch (Exception e)
+                {
+                    String className = testInfo.getTestClass().orElseThrow().getName();
+                    dumpHeap("server-" + className);
+                    fail(e.getMessage() + "\n---\nServer Leaks: " + serverBufferPool.dumpLeaks() + "---\n");
+                }
+            }
```
Comment (sbordet) [inline jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/AbstractTest.java]: Utility method.

File: jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/AbstractTest.java [hunk 15]
```diff
@@ -107,10 +121,131 @@ public static Collection<Transport> transportsTCP()
...
+        }
+        finally
+        {
+            LifeCycle.stop(client);
+            LifeCycle.stop(server);
+        }
+    }
+
+    private static boolean isLeakTrackingDisabled(TestInfo testInfo, String tagSubValue)
```
Comment (sbordet) [inline jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/AbstractTest.java:154]: Ok for now, but eventually this method should go away.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java (hunk 8 above)
Comment (lorban) [reply jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java]: Ok, done.

File: jetty-core/jetty-fcgi/jetty-fcgi-server/pom.xml (hunk 9 above)
Comment (lorban) [reply jetty-core/jetty-fcgi/jetty-fcgi-server/pom.xml]: They aren't anymore, let me remove those.

File: jetty-core/jetty-fcgi/jetty-fcgi-server/src/test/java/org/eclipse/jetty/fcgi/server/AbstractHttpClientServerTest.java (hunk 10 above)
Comment (lorban) [reply jetty-core/jetty-fcgi/jetty-fcgi-server/src/test/java/org/eclipse/jetty/fcgi/server/AbstractHttpClientServerTest.java]: Done.

File: jetty-core/jetty-fcgi/jetty-fcgi-server/src/test/java/org/eclipse/jetty/fcgi/server/HttpClientTest.java (hunk 11 above)
Comment (lorban) [reply jetty-core/jetty-fcgi/jetty-fcgi-server/src/test/java/org/eclipse/jetty/fcgi/server/HttpClientTest.java]: Done.

File: jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/DebugHandlerTest.java (hunk 12 above)
Comment (lorban) [reply jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/DebugHandlerTest.java]: Done.

File: jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ssl/ServerConnectorSslServerTest.java (hunk 13 above)
Comment (lorban) [reply jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ssl/ServerConnectorSslServerTest.java]: Done.

File: jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/AbstractTest.java (hunk 14 above)
Comment (lorban) [reply jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/AbstractTest.java]: Done.

File: jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/AbstractTest.java (hunk 15 above)
Comment (lorban) [reply jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/AbstractTest.java:154]: Agreed, once all the leaks are gone, we won't need this anymore.

**Review by sbordet — APPROVED** [review]

**Review by gregw — APPROVED** [review]
Looks good.... but should `dumpLeaks` be a JMX accessible method?

**Comment by lorban** [comment]
> Looks good.... but should `dumpLeaks` be a JMX accessible method?

I don't think exposing `dumpLeaks` via JMX would provide much value as you need to ensure Jetty is at rest before you call it, otherwise you get false positives.

**Comment by lorban** [comment]
I can't manage to get a green build but the last few builds has no assertion failure because of a leak.
